### PR TITLE
ddl: fix a bug that cancel add index ddl job when workers are not run, which cause tidb ddl hang up

### PR DESCRIPTION
**Pull request by winkyao**
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

If we cancel add index ddl job when the `job.Type == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer == 0`, the TiDB ddl will hang up like:

```
MySQL [(none)]> admin show ddl jobs;                                                                                                                                                   [11/973]
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+                                           
| JOBS                                                                                                                                                                                                                                | STATE     |                                           
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------+                                           
| ID:230, Type:drop index, State:running, SchemaState:none, SchemaID:31, TableID:48, RowCount:0, ArgLen:0, start time: 2018-10-25 17:24:31.123 +0800 CST, Err:[ddl:101]invalid table state public, ErrCount:344905, SnapshotVersion:0 | running   |                                           
| ID:231, Type:drop index, State:none, SchemaState:none, SchemaID:31, TableID:51, RowCount:0, ArgLen:0, start time: 2018-10-25 17:27:10.173 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0                                       | none      |                                           
| ID:232, Type:drop index, State:none, SchemaState:none, SchemaID:31, TableID:54, RowCount:0, ArgLen:0, start time: 2018-10-25 17:27:11.273 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0                                       | none      |                                           
```

### What is changed and how it works?

```
// runDDLJob runs a DDL job. It returns the current schema version in this transaction and the error.
func (w *worker) runDDLJob(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
   log.Infof("[ddl-%s] run DDL job %s", w, job)
   timeStart := time.Now()
   defer func() {
      metrics.DDLWorkerHistogram.WithLabelValues(metrics.WorkerRunDDLJob, job.Type.String(), metrics.RetLabel(err)).Observe(time.Since(timeStart).Seconds())
   }()
   if job.IsFinished() {
      return
   }
   // The cause of this job state is that the job is cancelled by client.
   if job.IsCancelling() {
      // If the value of SnapshotVer isn't zero, it means the work is backfilling the indexes.
      if job.Type == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer != 0 {
         log.Infof("[ddl-%s] run the cancelling DDL job %s", w, job)
         w.reorgCtx.notifyReorgCancel()
      } else {
         job.State = model.JobStateCancelled
         job.Error = errCancelledDDLJob
         job.ErrorCount++
         return
      }
   }
```

if `job.Type == model.ActionAddIndex && job.SchemaState == model.StateWriteReorganization && job.SnapshotVer == 0`, we will set job.State to model.JobStateCancelled, unfortunately, the adding index state is model.StateWriteReorganization, and it will not be rollbacked. So when we execute drop index, we will got error:

```
tidb-2018-10-25T19-14-23.472.log:2018/10/25 17:44:44.356 ddl_worker.go:329: [info] [ddl] run DDL job ID:230, Type:drop index, State:running, SchemaState:none, SchemaID:31, TableID:48, RowCount:0, ArgLen:0, start time: 2018-10-25 17:24:31.123 +0800 CST, Err:[ddl:101]invalid table state write reorganization, ErrCount:172588, SnapshotVersion:0
```
it is returned by:

```
func onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
	...

	originalState := indexInfo.State
	switch indexInfo.State {
	case model.StatePublic:
                ...
	case model.StateWriteOnly:
		...
	case model.StateDeleteOnly:
		...
	case model.StateDeleteReorganization:
		...
	default:
		err = ErrInvalidTableState.GenWithStack("invalid table state %v", tblInfo.State)
	}
```


### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test


Related changes

 - Need to cherry-pick to the release branch

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/8171)
<!-- Reviewable:end -->

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -437,12 +437,13 @@ func onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 		if job.IsRollingback() {
 			job.FinishTableJob(model.JobStateRollbackDone, model.StateNone, ver, tblInfo)
 			job.Args[0] = indexInfo.ID
+			job.Args = append(job.Args, getPartitionIDs(tblInfo))
```
Comment (ciscoxll) [inline ddl/index.go]: Is there a compatibility issue here?

**Comment by winkyao** [comment]
@ciscoxll @zimulala PTAL

File: ddl/db_test.go [hunk 2]
```diff
@@ -455,6 +456,65 @@ LOOP:
...
+		s.mustExec(c, "insert into t values (?, ?)", i, i)
+	}
+
+	var checkErr error
+	oldReorgWaitTimeout := ddl.ReorgWaitTimeout
+	ddl.ReorgWaitTimeout = 50 * time.Millisecond
+	defer func() { ddl.ReorgWaitTimeout = oldReorgWaitTimeout }()
+	hook := &ddl.TestDDLCallback{}
+	hook.OnJobRunBeforeExported = func(job *model.Job) {
+		log.Infof("hook.OnJobRunBeforeExported, job: %v", job)
```
Comment (zimulala) [inline ddl/db_test.go]: Remove this debug log?

File: ddl/db_test.go [hunk 3]
```diff
@@ -455,6 +456,65 @@ LOOP:
 	s.dom.DDL().(ddl.DDLForTest).SetHook(callback)
 }
 
+func (s *testDBSuite) TestCancelAddIndex1(c *C) {
```
Comment (zimulala) [inline ddl/db_test.go]: Please add some comments to this test.

**Comment by winkyao** [comment]
I'll add more test cases of canceling add column later.

File: ddl/rollingback.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
+// Copyright 2015 PingCAP, Inc.
```
Comment (crazycs520) [inline ddl/rollingback.go]: s/2015/2018

File: ddl/rollingback.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	err = job.DecodeArgs(&unique, &indexName, &idxColNames, &indexOption)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	indexInfo := findIndexByName(indexName.L, tblInfo.Indices)
+	if indexInfo == nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Errorf("index not exist %s", indexName)
```
Comment (crazycs520) [inline ddl/rollingback.go]: What if the job is canceled after the job EnQueueDDLJob, but before the owner first run the job?  Then will return error "index not exist" to client, but should return "cancelled DDL job".

File: ddl/rollingback.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	tblInfo, err := getTableInfo(t, job, schemaID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	columnInfo := model.FindColumnInfo(tblInfo.Columns, col.Name.L)
+	if columnInfo == nil {
+		job.State = model.JobStateCancelled
+		return ver, infoschema.ErrColumnNotExists.GenWithStackByArgs(col.Name)
```
Comment (crazycs520) [inline ddl/rollingback.go]: error miss the table name.
infoschema.ErrColumnNotExists.GenWithStackByArgs(col.Name, tblInfo.Name)

File: ddl/rollingback.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfo(t, job, schemaID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	columnInfo := model.FindColumnInfo(tblInfo.Columns, col.Name.L)
+	if columnInfo == nil {
```
Comment (crazycs520) [inline ddl/rollingback.go]: ditto with line #73.

File: ddl/rollingback.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(errCancelledDDLJob)
```
Comment (crazycs520) [inline ddl/rollingback.go]: return ver, errCancelledDDLJob

File: ddl/rollingback.go [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+	err = job.DecodeArgs(&unique, &indexName, &idxColNames, &indexOption)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	indexInfo := findIndexByName(indexName.L, tblInfo.Indices)
+	if indexInfo == nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(errCancelledDDLJob)
```
Comment (crazycs520) [inline ddl/rollingback.go]: return ver, errCancelledDDLJob

**Comment by crazycs520** [comment]
@zimulala @ciscoxll  PTAL

**Comment by winkyao** [comment]
@crazycs520 PTAL

File: ddl/rollingback.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+		return ver, errCancelledDDLJob
+	}
+	return convertAddIdxJob2RollbackJob(t, job, tblInfo, indexInfo, occuredErr)
+}
+
+func rollingbackAddColumn(t *meta.Meta, job *model.Job) (ver int64, err error) {
+	job.State = model.JobStateRollingback
+	col := &model.ColumnInfo{}
+	pos := &ast.ColumnPosition{}
+	offset := 0
```
Comment (tiancaiamao) [inline ddl/rollingback.go]: Should offet be math.MaxUint64 ?

File: ddl/db_test.go (hunk 3 above)
Comment (winkyao) [reply ddl/db_test.go]: Done

File: ddl/rollingback.go (hunk 10 above)
Comment (winkyao) [reply ddl/rollingback.go]: I don't think it is necessary. The initial offset value is the same with onAddColumn.

Comment (winkyao) [reply ddl/rollingback.go]: Actually, this param is useless but just decode it.

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/rollingback.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	// The write reorganization state in add index job that likes write only state in drop index job.
+	// So the next state is delete only state.
+	originalState := indexInfo.State
+	indexInfo.State = model.StateDeleteOnly
+	job.SchemaState = model.StateDeleteOnly
+	ver, err1 := updateVersionAndTableInfo(t, job, tblInfo, originalState != indexInfo.State)
+	if err1 != nil {
+		return ver, errors.Trace(err1)
+	}
+
```
Comment (crazycs520) [inline ddl/rollingback.go]: remove blank line.

**Pull request merged**